idea/frontend/src/features/dns/sails/index.ts

```diff
@@ -0,0 +1,114 @@
+import { TransactionBuilder } from 'sails-js';
+import { HexString } from '@gear-js/api';
+import { Program } from './lib';
+
+import { useEffect, useState } from 'react';
+import { useAccount, useAlert, useApi } from '@gear-js/react-hooks';
+import { web3FromSource } from '@polkadot/extension-dapp';
+import { DNS_API_URL, TransactionName } from '@/shared/config';
+import { useLoading } from '../hooks';
+import { useModal } from '@/hooks';
+
+let dnsProgram: Program;
+
+type ResolveRejectOptions = {
+  resolve?: () => void;
+  reject?: () => void;
+};
+
+type SendDnsMessage = {
+  getTransactionBuilder: () => Promise<TransactionBuilder<null>>;
+  options?: ResolveRejectOptions;
+};
+
+const useDnsActions = () => {
+  const [isLoading, enableLoading, disableLoading] = useLoading();
+  const { showModal } = useModal();
+  const { account } = useAccount();
+  const alert = useAlert();
+  const { api } = useApi();
+
+  const sendMessage = async ({ getTransactionBuilder, options }: SendDnsMessage) => {
+    if (!account || !api) {
+      return;
+    }
+    const { resolve: onSuccess, reject: onError } = options || {};
+    const { signer } = await web3FromSource(account.meta.source);
+    // ! TODO: calculate fee by sails
+    // const { partialFee } = await api.message.paymentInfo(account.address, { signer });
+
+    return new Promise<void>((resolve, reject) => {
+      const onConfirm = async () => {
+        if (account) {
+          try {
+            enableLoading();
+            const transaction = await getTransactionBuilder();
+
+            transaction.withAccount(account.address, { signer });
+
+            await transaction.calculateGas();
+            const result = await transaction.signAndSend();
```

Comment:
After sails update we will have an access to extrinsic. I suggest to take a look at `useSignAndSend` hook, maybe it would be possible to reuse it's logic.